lessons/06-tool-calling/README.md

```diff
@@ -0,0 +1,426 @@
+const openai = new OpenAI({
+    baseURL: "https://models.inference.ai.azure.com", // might need to change to this url in the future: https://models.github.ai/inference
+    apiKey: process.env.GITHUB_TOKEN,
+});
+
+// 5: Define the messages that will be sent to the AI model
+const messages = [
+{
+    role: "system",
+    content: `You are a helpful assistant. You can call functions to perform tasks. Make sure to parse the function call and arguments correctly.`
+}, {
+    role: "user",
+    content: "Find a landing spot given coordinates 8.5, 130.5"
+}
+];
+
+async function main(){
+  console.log("Making LLM call")
+
+  // 6: Call the AI model with the defined messages and tools
+  const result = await openai.chat.completions.create({
+      model: 'gpt-4o',
+      messages: messages,
+      functions: [findLandingSpotJson]
+    });
+
+  for (const choice of result.choices) {
+
+      let functionCall = choice.message?.function_call;
+      let functionName = functionCall?.name;
+      let args = JSON.parse(functionCall?.arguments);
+
+      // 7: Interpret response and call the tool based on the function call provided by the AI model
+      if (functionName && functionName in tools) {
+          console.log(`Calling [${functionName}]`);
+          const toolFunction = tools[functionName];
+          const toolResponse = toolFunction(...Object.values(args)); // Extract values from args and spread them
+          console.log("Result from [tool] calling: ", toolResponse);
+      }
+  }
+}
+
+main();
+```
+
+In the preceding code we've:
+
+- Created a function called `findLandingSpot` that takes latitude and longitude as input and returns the coordinates of a suitable landing spot.
+- Defined the metadata for the `find-landing-spot` tool.
+- Created a `tools` object that maps tool names to tool metadata.
+- Provided the `tools` object to the `client.getChatCompletions` call.
+
+   ```javascript
+   if (functionName && functionName in tools) {
+    console.log(`Calling [${functionName}]`);
+    const toolFunction = tools[functionName];
+    const toolResponse = toolFunction(...Object.values(args)); // Extract values from args and spread them
+    console.log("Result from [tool] calling: ", toolResponse);
+   }
+   ```
+
+- Called the tool based on the function call provided by the AI model.
+- Printed the result of the tool call.
+
+**You**: "I think I get it. I define a function, create a metadata representation of the tool, provide the metadata to the AI model, and then call the tool based on the function call provided by the AI model."
+
+**Time beetle**: "Exactly! You're ready to start building your own tools and integrating them with the AI model."
+
+## Assignment - upgrade Amelia's time travel device
+
+**Amelia Earhart**: "We're coming down hot, thank God you found us a landing spot. Hold on tight!"
+
+Amelia ends up expertly landing the plane on a small island. You and Amelia step out of the plane, whereby Amelia hands you a small device.
```

Comment:
I'd modify this sentence slightly:

Amelia expertly lands the plane on a small island.